kubernetes/client_factory.go

```diff
@@ -290,51 +312,82 @@ func getTokenHash(authInfo *api.AuthInfo) string {
 }
 
 // KialiSAClients returns all clients associated with the Kiali service account across clusters.
-func (cf *clientFactory) GetSAClients() map[string]ClientInterface {
-	cf.mutex.RLock()
-	defer cf.mutex.RUnlock()
-
-	for cluster := range cf.saClientEntries {
-		if err := cf.refreshClientIfTokenChanged(cluster); err != nil {
-			log.Errorf("Unable to refresh Kiali SA client for cluster: %s. Err: %s", cluster, err)
-		}
+func (cf *clientFactory) GetSAClient(cluster string) ClientInterface {
+	// while we are here, refresh the client
+	if err := cf.refreshClientIfTokenChanged(cluster); err != nil {
+		log.Errorf("Unable to refresh Kiali SA client for cluster [%s]: %v", cluster, err)
 	}
 
-	return cf.saClientEntries
+	cf.mutex.RLock()
+	defer cf.mutex.RUnlock()
+	return cf.saClientEntries[cluster]
 }
 
 // Check for kiali token changes and refresh the client when it does.
 func (cf *clientFactory) refreshClientIfTokenChanged(cluster string) error {
-	// TODO: This only checks that the token has changed. When the client factory supports reading
-	// the tokens from the filesystem, then this should check the client against the token on disk
-	// that is associated with the cluster.
-	kialiSAToken, err := GetKialiToken()
-	if err != nil {
-		return err
+	var refreshTheClient bool // will be true if the client needs to be refreshed
+	var rci *RemoteClusterInfo
+
+	if cluster == HomeClusterName {
+		// LOCAL CLUSTER
+		if newTokenToCheck, err := GetKialiTokenForHomeCluster(); err != nil {
+			return err
+		} else {
+			cf.mutex.RLock()
+			client, ok := cf.saClientEntries[cluster]
+			cf.mutex.RUnlock()
+			if !ok {
+				return fmt.Errorf("There is no home cluster SA client to refresh")
+			}
+			refreshTheClient = client.GetToken() != newTokenToCheck
+			rci = nil
+		}
+	} else {
+		// REMOTE CLUSTER
+		cf.mutex.RLock()
+		remoteRci, ok := cf.remoteClusterInfos[cluster]
+		cf.mutex.RUnlock()
+		if !ok {
+			return fmt.Errorf("Cannot refresh token for unknown cluster [%s]", cluster)
+		} else {
+			if reloadedRci, err := reloadRemoteClusterInfoFromFile(remoteRci); err != nil {
+				return err
+			} else {
+				refreshTheClient = (reloadedRci != nil) // note that anything (not just the token) that changed will trigger the client to be refreshed
+				rci = reloadedRci
+			}
+		}
 	}
 
-	if cf.saClientEntries[cluster].GetToken() != kialiSAToken {
-		log.Debugf("Kiali SA token has changed, refreshing client for cluster: %s", cluster)
-		// Token has changed, so we need to refresh the client.
-		newClient, err := cf.newSAClient(cluster)
+	if refreshTheClient {
+		log.Debugf("Kiali SA token has changed for cluster [%s], refreshing the client", cluster)
```

Comment:
I've noticed that for the local cluster, it always throws this message, but I don't think anything changes there, for some reason, the token from the client is "" so is always different from the token it uses to check.

Reply:
I see this about every 48 minutes (not exactly 48, but always around 48).
```
2023-02-10T14:35:33Z DBG Kiali SA token has changed for cluster [_kiali_home], refreshing the client
2023-02-10T15:23:41Z DBG Kiali SA token has changed for cluster [_kiali_home], refreshing the client
2023-02-10T16:11:50Z DBG Kiali SA token has changed for cluster [_kiali_home], refreshing the client
2023-02-10T17:01:00Z DBG Kiali SA token has changed for cluster [_kiali_home], refreshing the client
2023-02-10T17:50:13Z DBG Kiali SA token has changed for cluster [_kiali_home], refreshing the client
2023-02-10T18:38:21Z DBG Kiali SA token has changed for cluster [_kiali_home], refreshing the client
2023-02-10T19:28:30Z DBG Kiali SA token has changed for cluster [_kiali_home], refreshing the client
2023-02-10T20:16:39Z DBG Kiali SA token has changed for cluster [_kiali_home], refreshing the client
```

Reply:
In my enviroment (I don't have multi cluster or secrets created now), just single cluster, it's happening more often:

```
2023-02-10T17:33:16-03:00 DBG Kiali SA token has changed for cluster [_kiali_home], refreshing the client
2023-02-10T17:33:26-03:00 DBG Kiali SA token has changed for cluster [_kiali_home], refreshing the client
2023-02-10T17:33:36-03:00 DBG Kiali SA token has changed for cluster [_kiali_home], refreshing the client
```

Reply:
This is caused when the file `/var/run/secrets/kubernetes.io/serviceaccount/token` changes. If your token is always `""` then that means it can't read that file. I've seen this when I run the server via `run-kiali.sh` and there is no file with that absolute pathname.